network/discovery/discovery.go

```diff
@@ -0,0 +1,424 @@
+	peerDiscoveryTicker := time.NewTicker(peerDiscoveryInterval)
+	bootnodeDiscoveryTicker := time.NewTicker(bootnodeDiscoveryInterval)
+
+	defer func() {
+		peerDiscoveryTicker.Stop()
+		bootnodeDiscoveryTicker.Stop()
+	}()
+
+	for {
+		select {
+		case <-d.closeCh:
+			return
+		case <-peerDiscoveryTicker.C:
+			go d.regularPeerDiscovery()
+		case <-bootnodeDiscoveryTicker.C:
+			go d.bootnodePeerDiscovery()
+		}
+	}
+}
+
+// regularPeerDiscovery grabs a random peer from the list of
+// connected peers, and attempts to find / connect to their peer set
+func (d *DiscoveryService) regularPeerDiscovery() {
+	if !d.baseServer.HasFreeConnectionSlot(network.DirOutbound) {
+		// No need to do peer discovery if no open connection slots
+		// are available
+		return
+	}
+
+	// Grab a random peer from the base server's peer store
+	peerID := d.baseServer.GetRandomPeer()
+	if peerID == nil {
+		// The node cannot find a random peer to query
+		// from the current peer set
+		return
+	}
+
+	// Try to discover the peers connected to the reference peer
+	if err := d.attemptToFindPeers(*peerID); err != nil {
+		d.logger.Error(
+			"Failed to find new peers",
+			"peer",
+			peerID,
+			"err",
+			err,
+		)
+	}
+}
+
+// bootnodeDiscovery queries a random (unconnected) bootnode for new peers
+// and adds them to the routing table
+func (d *DiscoveryService) bootnodePeerDiscovery() {
+	if !d.baseServer.HasFreeConnectionSlot(network.DirOutbound) {
+		// No need to attempt bootnode dialing, since no
+		// open outbound slots are left
+		return
+	}
+
+	var (
+		isTemporaryDial bool           // dial status of the connection
+		bootnode        *peer.AddrInfo // the reference bootnode
+	)
+
+	// Try to find a suitable bootnode to use as a reference peer
+	for bootnode == nil {
+		// Get a random unconnected bootnode from the bootnode set
+		bootnode = d.baseServer.GetRandomBootnode()
+		if bootnode == nil {
+			// No bootnodes available
+			return
+		}
+
+		// If one or more bootnode is connected the dial status is temporary
+		if d.baseServer.GetBootnodeConnCount() > 0 {
+			// Check if the peer is already a temporary dial
+			if alreadyTempDial := d.baseServer.FetchOrSetTemporaryDial(
+				bootnode.ID,
+				true,
+			); alreadyTempDial {
+				bootnode = nil
+
+				continue
+			}
+
+			// Mark the subsequent connection as temporary
+			isTemporaryDial = true
+		}
+	}
+
+	defer func() {
+		if isTemporaryDial {
+			// Since temporary dials are short-lived, the connection
+			// needs to be turned off the moment it's not needed anymore
+			d.baseServer.RemoveTemporaryDial(bootnode.ID)
+			d.baseServer.DisconnectFromPeer(bootnode.ID, "Thank you")
+		}
+	}()
+
+	// Find peers from the referenced bootnode
+	foundNodes, err := d.findPeersCall(bootnode.ID, true)
+	if err != nil {
+		d.logger.Error("Unable to execute bootnode peer discovery, %w", err)
+
+		return
+	}
+
+	// Save the peers for subsequent dialing
+	d.addPeersToTable(foundNodes)
```

Comment:
I faced error when server tries to connect one of bootnodes, which is not connected to server now

```
2022-03-31T15:40:47.937+0300 [ERROR] polygon.network.discovery: Unable to execute bootnode peer discovery, %w: EXTRA_VALUE_AT_END="unable to create new discovery client connection, peer disconnected before the discovery client was initialized"
```

Reply:
You're right, temporary dials **are never added to the peer store**, since they are temporary (only for a few moments).

Fixed the Discovery client method to reflect this 👍 :

[449cd90](https://github.com/0xPolygon/polygon-edge/pull/343/commits/449cd90d57247c3b5f8501094fba2dc632af0e2a)